### ISSUE TITLE
 variable 'ClassName' used before declaration

### ISSUE THREAD

**Issue by fkolar** [palantir/tslint#3565]
### Bug Report

- __TSLint version__: 5.7.0
- __TypeScript version__: 2.4.2
- __Running TSLint via__: (pick one) CLI 

#### TypeScript code being linted

```ts
import {Component, OnInit} from "@angular/core";

@Component({
  selector: "app-root",
  templateUrl: "./app.component.html",
  styleUrls: ["./app.component.css"]
})
export class AppComponent implements OnInit {
  title = "app";


  ngOnInit(): void {
    let test = new MyTestClass();

    console.log(test);
  }
}


export class MyTestClass {

  constructor() {
  }
}
```

with `tslint.json` configuration:

- Default one that is created during `ng new` process

#### Actual behavior

I can not define a class in the same file whihc I see inside @angular code base pretty normal thing

#### Expected behavior

This should not be an error. I could understand use this for actual variables let,const but not really for a types.

**Comment by JoshuaKGoldberg** [comment]
@fkolar you're using `MyTestClass` inside the `AppComponent` class, but `AppComponent` is declared before `MyTestClass`. If you move `MyTestClass` to before `AppComponent` does the error go away? It should.

**Comment by fkolar** [comment]
Yes it goes away, but what I would expect this would not be an error. Typescript compiler does not have problem with this, NG code uses  the same style all over or just like for any other lang its normal things. 

`MyTestClass` is declared in the same file not really inside `AppComponent` .

**Comment by JoshuaKGoldberg** [comment]
It would be difficult to draw the line between safe cases like what you proposed and code that does use them before instantiation:

```typescript
class First {
    public readonly second = new Second();
}

class Second {
    public static readonly first = new First();
}
```

Under the hood, classes often compile to something that does actually use `var`s like:

```javascript
var MyTestClass = /** @class */ (function () {
    function MyTestClass() {
    }
    return MyTestClass;
}());
```

**Comment by puvaasam** [comment]
What will happen if i keep those class in different files

file-name:  my-test-class.ts -> 
class MyTestClass {
   
}

file-name: app.component.ts -> 

class AppComponent {
     public ngOnint(){
        new MyTestClass ();
    }
}

**Comment by JoshuaKGoldberg** [comment]
@puvaasam That's very fine, and works with this rule. See http://www.typescriptlang.org/docs/handbook/modules.html

**Comment by fkolar** [comment]
Thanks for the feedback.  Its seem its by design.
